### PR TITLE
wr-169: add-edit-article

### PR DESCRIPTION
**Pull request by canterbery**
image
image

### CODE REVIEW

File: backend/src/packages/articles/article.service.ts [hunk 1]
```diff
@@ -139,14 +143,20 @@ class ArticleService implements IService {
       });
     }
 
+    if (article.userId !== user.id) {
+      throw new ApplicationError({
+        message: 'Article can only be edited by author!',
+      });
```
Comment (andanf-e) [inline backend/src/packages/articles/article.service.ts]:
```suggestion
      throw new ForbiddenError('Article can only be edited by author!');
```

File: frontend/src/pages/articles/components/article-card/article-card.tsx [hunk 2]
```diff
@@ -52,7 +61,17 @@ const ArticleCard: React.FC<Properties> = ({
           )}
           <span className={styles.publicationTime}>{MOCKED_READ_TIME}</span>
         </div>
-        <Icon iconName="favorite" className={styles.icon} />
+        <div className={styles.iconWrapper}>
+          {isMyArticles && (
+            <RouterLink
+              to={`${AppRoute.EDIT_ARTICLE_BASE}${id}`}
```
Comment (andanf-e) [inline frontend/src/pages/articles/components/article-card/article-card.tsx]:
```suggestion
              to={AppRoute.EDIT_ARTICLE.replace(':id', id)}
```

File: frontend/src/libs/enums/app-route.enum.ts [hunk 3]
```diff
@@ -10,6 +10,8 @@ const AppRoute = {
   CREATE_ARTICLE: '/articles/create',
   FORGOT_PASSWORD: '/forgot-password',
   RESET_PASSWORD: RESET_PASSWORD_ROUTE.WITH_PARAM,
+  EDIT_ARTICLE: '/articles/edit/:id',
+  EDIT_ARTICLE_BASE: '/articles/edit/',
```
Comment (andanf-e) [inline frontend/src/libs/enums/app-route.enum.ts]: Not sure if we need this
```suggestion

```

File: frontend/src/pages/article/components/article-form/article-form.tsx [hunk 4]
```diff
@@ -1,15 +1,19 @@
+import { articleUpdateValidationSchema } from 'shared/build/index.js';
```
Comment (andanf-e) [inline frontend/src/pages/article/components/article-form/article-form.tsx]:
```suggestion
import { articleUpdateValidationSchema } from '~/packages/articles/articles.js';
```

File: frontend/src/slices/articles/actions.ts [hunk 5]
```diff
@@ -56,6 +60,21 @@ const createArticle = createAsyncThunk<
...
+  ArticleUpdateRequestPayload,
+  AsyncThunkConfig
+>(`${sliceName}/update`, async (payload, { extra }) => {
+  const { articleApi } = extra;
+
+  const article = await articleApi.update(payload);
+  const { navigate } = payload;
+  if (navigate) {
+    navigate(`${AppRoute.ARTICLES}/${ArticleSubRoute.MY_ARTICLES}`);
+  }
```
Comment (andanf-e) [inline frontend/src/slices/articles/actions.ts]:
```suggestion

```

File: frontend/src/pages/article/components/article-form/article-form.tsx [hunk 6]
```diff
@@ -50,19 +67,44 @@ const ArticleForm: React.FC = () => {
...
 
+  const handleArticleUpdate = useCallback(
+    (payload: ArticleRequestDto): void => {
+      if (!articleForUpdate) {
+        return;
+      }
+      const updatePayload = {
+        articleId: articleForUpdate.id,
+        articleForUpdate: { text: payload.text, title: payload.title },
+        navigate,
```
Comment (andanf-e) [inline frontend/src/pages/article/components/article-form/article-form.tsx]:
```suggestion

```

File: frontend/src/pages/article/components/article-form/article-form.tsx [hunk 7]
```diff
@@ -50,19 +67,44 @@ const ArticleForm: React.FC = () => {
...
+      };
+      void dispatch(articlesActions.updateArticle(updatePayload));
```
Comment (andanf-e) [inline frontend/src/pages/article/components/article-form/article-form.tsx]:
```suggestion
      void dispatch(articlesActions.updateArticle(updatePayload))
        .unwrap()
        .then(() => {
          navigate
        });
```

File: frontend/src/pages/article/edit-article-page.tsx [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+import { ArticleForm } from './components/components.js';
+import styles from './styles.module.scss';
+
+const EditArticlePage: React.FC = () => {
+  const location = useLocation();
+  return (
+    <Layout>
+      <div className={styles.articleCreatePageWrapper}>
+        <ArticleForm
+          articleForUpdate={location.state as ArticleWithAuthorType}
```
Comment (andanf-e) [inline frontend/src/pages/article/edit-article-page.tsx:20]: Maybe let's check if location.state is not empty before passing it

File: frontend/src/slices/articles/articles.slice.ts [hunk 9]
```diff
@@ -27,6 +33,16 @@ const { reducer, actions, name } = createSlice({
       state.articles = [...state.articles, action.payload];
       state.dataStatus = DataStatus.FULFILLED;
     });
+    builder.addCase(updateArticle.fulfilled, (state, action) => {
+      const article = action.payload;
+      if (article) {
+        state.articles = [
+          article,
+          ...state.articles.filter((item) => item.id !== article.id),
+        ];
```
Comment (andanf-e) [inline frontend/src/slices/articles/articles.slice.ts]:
```suggestion
        state.articles = state.articles.map(item => {
          if (article.id = item.id) {
            return article;
          }
          return item;
        });
```

File: shared/src/packages/articles/libs/types/article-update-request-payload.type.ts [hunk 10]
```diff
@@ -0,0 +1,11 @@
+import { type NavigateFunction } from 'react-router-dom';
+
+import { type ArticleUpdateRequestDto } from './article-update-request-dto.type.js';
+
+type ArticleUpdateRequestPayload = {
+  articleId: number;
+  articleForUpdate: ArticleUpdateRequestDto;
+  navigate?: NavigateFunction;
```
Comment (andanf-e) [inline shared/src/packages/articles/libs/types/article-update-request-payload.type.ts]:
```suggestion

```

**Review by andanf-e — CHANGES_REQUESTED** [review]

File: frontend/src/pages/article/components/article-form/article-form.tsx (hunk 7 above)
Comment (canterbery) [reply frontend/src/pages/article/components/article-form/article-form.tsx]: isn't unwrap will throw errors as well if action is rejected ?

File: frontend/src/pages/article/edit-article-page.tsx (hunk 8 above)
Comment (canterbery) [reply frontend/src/pages/article/edit-article-page.tsx:20]: And what I should do if it's empty ? return empty div ?

Comment (andanf-e) [reply frontend/src/pages/article/edit-article-page.tsx:20]: Redirect back to my-articles

File: frontend/src/pages/article/components/article-form/article-form.tsx (hunk 7 above)
Comment (canterbery) [reply frontend/src/pages/article/components/article-form/article-form.tsx]: image

This is unwrap. Should I leave it like that ? Or add some empty catch ?

File: frontend/src/pages/article/edit-article-page.tsx [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+import { type ArticleWithAuthorType } from '~/packages/articles/articles.js';
+
+import { ArticleForm } from './components/components.js';
+import styles from './styles.module.scss';
+
+const EditArticlePage: React.FC = () => {
+  const location = useLocation();
+  const navigate = useNavigate();
+  if (!location.state) {
+    navigate(`${AppRoute.ARTICLES}/${ArticleSubRoute.MY_ARTICLES}`);
```
Comment (andanf-e) [inline frontend/src/pages/article/edit-article-page.tsx]:
```suggestion
  return <Navigate to={`${AppRoute.ARTICLES}/${ArticleSubRoute.MY_ARTICLES}`} />;
```

**Review by andanf-e — DISMISSED** [review]

**Review by mhrechkin — DISMISSED** [review]

**Review by NazarKovtun1 — DISMISSED** [review]

**Review by andanf-e — APPROVED** [review]

**Pull request merged**